### ISSUE TITLE
Reload configuration not working

### ISSUE THREAD

**Issue by marke72** [elastic/logstash#6417]
Hello,

I recently updated from 5.0.2 to 5.1.1 of logstash. I have been using the reload configuration feature for a while now but it seems that in 5.1.1 it has stopped working. Logstash acknowledges in the logs that it received a SIGHUP
`[2016-12-14T08:03:00,226][WARN ][logstash.runner          ] SIGHUP received.`
but it doesn't seem to actually reload the configuration. If I restart the service then it loads the new configuration correctly. I just tried doing a simple output to stdout, `stdout { codec => rubydebug }`, then reload the configuration but nothing is outputted to `/var/log/messages`.
I have used the reload configuration feature consistently since it's release. But with the latest Logstash release, 5.1.1, it seems to be broken.

Thanks in advanced!

- Version: 5.1.1 via yum repository
- Operating System: CentOS 7

**Comment by jsvd** [comment]
Confirmed @marke72, investigating.

**Comment by jsvd** [comment]
@marke72 in the meantime, is using -r an option as a workaround?

**Comment by marke72** [comment]
The auto reload does seem to work still. I can probably make it work for my environment. Thanks!

**Comment by jordansissel** [comment]
@marke72 Apologies for what is maybe a silly question, but when you send SIGHUP, has the logstash configuration file(s) changed content from when Logstash started?

**Comment by marke72** [comment]
Yeah. I run Logstash as a service. The service will already be started and running just fine. If I need to fix or change something I'll make the changes in the configuration file located in the `/etc/logstash/conf.d` directory. After saving the file I'll run a test on the file with the `-t` option and if it passes I would send a SIGHUP to the logstash process created by the service. Only since updating Logstash to 5.1.1 from 5.0.2 has it stopped working.

**Comment by dav3860** [comment]
Same behavior here. The SIGHUP is received, as it appears in the logs :
`[2016-12-20T16:35:41,839][WARN ][logstash.runner          ] SIGHUP received.`

But logstash does not reload the configuration.

**Comment by imweijh** [comment]
same here.
send HUP : 
# kill -HUP $(pgrep -G logstash)

log read as:
[2016-12-22T20:39:10,611][WARN ][logstash.runner          ] SIGHUP received.

**Comment by ZhukovAlexey** [comment]
Have the same issue using 5.1.1 with Docker.

After sendng kill -1 in logstash log there's a line
`[2016-12-23T17:21:41,355][WARN ][logstash.runner          ] SIGHUP received.`

But reload configuration doesn't happen.

**Comment by breml** [comment]
Just two additional infos:

* Logstash 5.2.x is also affected by this bug
* If Logstash is started with `-r`, the reload with SIGHUP works (tested with `--config.reload.interval <large-number>`)

**Comment by Shaoranlaos** [comment]
I'm no ruby programmer so I don't realy have the skill to fix this my self but i think the breaking point here is the line:
`@pipelines.each do |pipeline_id, pipeline|
        next if pipeline.settings.get("config.reload.automatic") == false
        begin
          reload_pipeline!(pipeline_id)`
([https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/agent.rb#L109](url))
(call of this function here: [https://github.com/elastic/logstash/blob/88eb425b00e1b41810ff924f3a0370ce51e919b2/logstash-core/lib/logstash/runner.rb#L388](url)

The function is called when an SIGHUP is received but it will skip the reload when the `config.reload.automatic` setting is not set.

Can someone with ruby skill fix this with this info?

**Comment by breml** [comment]
Links in the above comment are broken, so here the working links:
* https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/agent.rb#L109
* https://github.com/elastic/logstash/blob/88eb425b00e1b41810ff924f3a0370ce51e919b2/logstash-core/lib/logstash/runner.rb#L388

**Comment by breml** [comment]
@jsvd I had a look at this issue and the reverenced code. In https://github.com/elastic/logstash/pull/6093/files you changed the behavior for the reload by adding an additional check, if `pipeline.settings.get("config.reload.automatic") == false` in `reload_state!` (https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/agent.rb#L106). The reason for this change is ticket #6094 filed by you where you state, that even 
> if globally `config.reload.automatic` is set to true, it should still be possible for a pipeline to be registered without reloading on and not be reloaded. 

Could you provide some more details about the reason, why it should be possible to have pipelines with the automatic config reload disabled, even when the user sets this option globally to true?

I guess these reasons could also influence, how to solve this issue, because the question I was asking my self is, if the user sends a SIGHUB to Logstash, should the same rules apply as for the automatic reload or is this an other case? Or in other words, should a SIGHUB lead to all pipelines be reloaded or should SIGHUB also respect if automatic reload is disabled on a pipeline?

**Comment by ZhukovAlexey** [comment]
The documentation for LS 5.2 clearly states that 

"If Logstash is already running without auto-reload enabled, you can force Logstash to reload the config file and restart the pipeline by sending a SIGHUP (signal hangup) to the process running Logstash. "

Doesn't it?

**Comment by breml** [comment]
@ZhukovAlexey No it does not. It is a confirmed bug (see https://github.com/elastic/logstash/issues/6417#issuecomment-269000424). The question is, what is the right way to fix it.

**Comment by abraxxa** [comment]
Logstash 5.3.0 is also affected and the docs don't reflect this although it's the case since two versions.

**Comment by jsvd** [comment]
this will be fixed with https://github.com/elastic/logstash/pull/6632 provides a much better foundation for configuration loading and reloading. this PR targets master, so 6.0
There's no current fix on the 5.x.
As a reminder, using -r will provide configuration reloading driven by file changes, instead of manually triggered

**Comment by breml** [comment]
@jsvd to be honest, this is a quite disappointing way of handling this issue. As already pointed out in several comments, the function of this feature is clearly documented since 5.x and it got broken due to a new introduced feature. This is clearly a regression bug and at least in my opinion it is good taste to get such bugs fixed in the current stable release. (just my 2cts)

**Comment by jsvd** [comment]
@breml I agree this should be fixed for 5.x, I just haven't gotten the chance to fix it yet. The fix in 6.0 was a side effect from a refactor that happen, not really a "let's forget about 5.x"
I'll try to fix this soon during the 5.3 timeline and backport to 5.2

**Comment by mrjameshamilton** [comment]
just found the same problem. it should be fixed for 5.x to work as documented.

**Comment by driverpt** [comment]
Fixed in #7853

**Comment by jsvd** [comment]
The fix will be out in 5.6.0. If we ever end up doing a 5.5.2, it will be there too.

**Comment by driverpt** [comment]
@jsvd , is this fix Live ?

**Comment by saggu** [comment]
Did anyone make this work in the docker version of logstash 5.5.2 ?
Its not working for me

**Comment by driverpt** [comment]
@saggu , are you running with Auto Reload on ?

**Comment by saggu** [comment]
@driverpt thanks for the update.
I have auto reload on, will test again when that PR is merged.
We need the logstash docker as part of a docker compose to auto reload, with no manual intervention

**Comment by jpazsedano** [comment]
Using Logstash 7.1.0 via the official dockerfile and this issue still happens. Logstash shows in the logs that it has received the SIGHUP, but does nothing.

Also, it ignores the LS_OPTS environment variable, but I think that's a different issue.

**Comment by ricferr** [comment]
> Using Logstash 7.1.0 via the official docker image and this issue still happens. Logstash shows in the logs that it has received the SIGHUP, but does nothing.
> 
> Also, it ignores the LS_OPTS environment variable, but I think that's a different issue.

I can confirm I'm also seeing this issue with 7.4.2 installed as a .deb package on Debian GNU/Linux 8.10.
Yes, I know thats an ancient version of Debian.

**Comment by sgurjeet** [comment]
Seeing the same issue with logstash 7.8.0 official docker image.
config.reload.automatic: true
config.reload.interval: 300
is added to logstash.yml and yet it still doesn't reload the configuration after the set interval.

**Comment by DeoLeung** [comment]
> Using Logstash 7.1.0 via the official docker image and this issue still happens. Logstash shows in the logs that it has received the SIGHUP, but does nothing.
> 
> Also, it ignores the LS_OPTS environment variable, but I think that's a different issue.

the following does not work with `7.9.2`
```bash
docker-compose kill -s HUP logstash
```

**Comment by douph1** [comment]
Same here : logstash 7.13.2 
kill -SIGHUP <pid>
>[2021-06-25T11:52:15,935][WARN ][logstash.runner          ] SIGHUP received.

but no reload triggered

**Comment by WallnerAndreas** [comment]
I struggled with this and didn't understand why my logstash.yml didn't reload and was about to contact elastic. Then I realised that the documentation is confusing, and what they probably referring to is just pipeline config here when they say config: https://www.elastic.co/guide/en/logstash/current/reloading-config.html
And at this point I have no idea if half of the people here are trying to reload the logstash.yml config or just the pipeline config. It will NOT reload the entire config, just pipeline config.
Hope this message will help someone to not waste their time as I have!
